### ISSUE TITLE
Feature Request - Return final loss from tf.estimator.Estimator.train along with self.

### ISSUE THREAD

**Issue by GrandathePanda** [tensorflow/tensorflow#12093]
### Describe the problem
Things like hyper opt need to know the loss so that it can effectively pick the best hyper parameters for the model. Right now self is just returned, it is a one line change to the code since the loss is set the exact line before. I would be happy to initiate the PR myself assuming its wanted.

### Source code / logs
From:
```
loss = self._train_model(input_fn=input_fn, hooks=hooks)
logging.info('Loss for final step: %s.', loss)
return self
```

To:
```
loss = self._train_model(input_fn=input_fn, hooks=hooks)
logging.info('Loss for final step: %s.', loss)
return self, loss
```

**Comment by terrytangyuan** [comment]
I had a PR that might address your need as well: https://github.com/tensorflow/tensorflow/pull/12026

**Comment by GrandathePanda** [comment]
Just checked it out, I suppose I could pull it from the written summary after the training was done, definitely a viable solution. It would still be easier to just grab it from the return of the function, but I understand that changing the signature of the function may not be what 's best for the project in terms of breaking peoples already written code. I'd like to see some more discussion on this. If it turns out that most of the people who chime would prefer not to alter the signature then by all means I'd wait for yours to be merged in or even help out if anything needs doing. If people do like it we could also just add onto that pr.

**Comment by GrandathePanda** [comment]
So my work around for this at the moment is that I wrote a function that monkeypatches the function in the estimator at execution and returns the loss. Not pretty but it works for my purposes. I'd still like to see people's opinions on this. It seems completely reasonable to me for this function to return the final loss.

**Comment by tatatodd** [comment]
@jhseu Can you comment?  Thanks!

**Comment by terrytangyuan** [comment]
You can add a custom session run hook with something like following:

```
def after_run(self, context, values):
  losses = context.session.graph.get_collection("losses")
  print(context.session.run(losses))
```
and then attach this to your estimator.

**Comment by terrytangyuan** [comment]
Yeah this will be a huge compatibility break. @ispirmustafa Any thoughts on this since this is related to the PR I created/closed?

**Comment by ispirmustafa** [comment]
final training loss shows the loss on final mini batch. We do hyper parameter tuning by using result of evaluate not train.

**Comment by GrandathePanda** [comment]
Yup I was aware of that. Assuming loss has been steadily decreasing with no abnormality the final minibatch loss should still be just as useful for tuning assuming either overfitting is not an issue or you don't care about over fitting, which in my case is the latter. What I'm hearing though is that there is an easier way through an alternate method. I'd honestly forgotten I opened this suggestion till I just checked my alerts on here for another lib, my life hasn't been deeply halted by this and for that reason I'm going to close this since I just expect it to collect dust.

**Comment by OluwoleOyetoke** [comment]
@GrandathePanda 
Please I am having the same issue. I'm quite surprised at how its so difficult to get my loss value out of the TF Estimator object. I have asked the question on stack overflow also, still awaiting answers.

Please can you let me know the work around you used?

**Comment by GrandathePanda** [comment]
Hey so while I monkey patched the original method to suit my purposes in that instance. I later found it was simpler (and more correct) to just grab loss on my evaluation passes. The .evaluate() method on the estimator returns a dictionary of metrics you can specify in your model function, in the estimator spec, https://www.tensorflow.org/api_docs/python/tf/estimator/EstimatorSpec eval_metric_ops

**Comment by OluwoleOyetoke** [comment]
Thanks, @GrandathePanda , it worked!

**Comment by shaunstarbuckcivicom** [comment]
What function did you use with eval_metric_ops to get the loss?

**Comment by jagjeet-singh** [comment]
@GrandathePanda 
EstimatorSpec will return validation loss. Any idea how I can extract the training loss?

**Comment by EvanSchalton** [comment]
> ### Describe the problem
> Things like hyper opt need to know the loss so that it can effectively pick the best hyper parameters for the model. Right now self is just returned, it is a one line change to the code since the loss is set the exact line before. I would be happy to initiate the PR myself assuming its wanted.
> 
> ### Source code / logs
> From:
> 
> ```
> loss = self._train_model(input_fn=input_fn, hooks=hooks)
> logging.info('Loss for final step: %s.', loss)
> return self
> ```
> To:
> 
> ```
> loss = self._train_model(input_fn=input_fn, hooks=hooks)
> logging.info('Loss for final step: %s.', loss)
> return self, loss
> ```

If there is appetite for this change a hacky solution would be to add a parameter ``return_loss = False`` then use
```
if return_loss:
  return self,loss
return self
```

but that might be equally hacky
